File: crates/lanekeep-rules/rules/duplicate-implementation.ts
import { defineRule } from 'lanekeep'

/**
 * One implementation written twice.
 *
 * An agent that cannot see the whole corpus reimplements a helper that already exists, in
 * every individual file the result looks fine, and no per-file rule can notice. This rule
 * fingerprints every function body — identifiers and literal values erased, computed
 * host-side by `ctx.structureFingerprint` — and reports any two bodies with the same shape.
 *
 * Matching is by structure, not by text: renaming an identifier or changing a literal value
 * still matches (that is the point), while changing an operator or adding a statement does
 * not. Comments never matter — the fingerprint is computed over the body with comments
 * erased, so in TypeScript a with-docstring/without-docstring pair has the same shape and is
 * flagged like any other pair. A *python* docstring is a statement rather than a comment, so
 * there the two directions pull apart: differing docstrings still group (the string's text is
 * erased), while with-docstring against without differs by a statement and does not.
 *
 * One rule, five grammars: `language` names them and `query` carries one entry per grammar.
 * Grouping never crosses a language — parallel implementations in two languages have
 * different interior node kinds, so their bodies cannot share a fingerprint.
 *
 * The `minNodes` threshold keeps tiny bodies (getters, one-line callbacks) out: a default of
 * 40 fires on real helpers, not on every two-line pair.
 *
 * Like every cross-file rule, it is skipped under `--since`/`--staged` with a notice on
 * stderr (docs/architecture.md §8.4) — it reports on full runs only.
 *
 * @example
 * ```ts
 * import duplicateImplementation from 'lanekeep/duplicate-implementation'
 *
 * export default defineConfig({
 *   rules: [duplicateImplementation({ minNodes: 60 })],
 * })
 * ```
 */
export default function duplicateImplementation(options) {
  // The ignored-options trap: options reach a rule only by being closed over. A handler is
  // invoked with two arguments; a third parameter would be `undefined` on every call.
  const minNodes = options?.minNodes ?? 40

  return defineRule({
    id: 'lanekeep/duplicate-implementation',
    language: ['typescript', 'tsx', 'python', 'go', 'rust'],
    severity: 'error',

    card: {
      message: 'duplicated implementation',
      remediation:
        'delete one copy and import the other — two bodies with one shape are one change that has to happen twice',
      examples: {
        bad: 'function normalize(raw) { return raw.trim().toLowerCase() }\nfunction clean(input) { return input.trim().toLowerCase() }',
        good: 'import { normalize } from "./normalize"',
      },
    },

    // The whole corpus feeds one pass, so the default per-invocation budget is the wrong
    // shape for it. See docs/cross-file-rules.md.
    timeout: 5_000,

    // One query per grammar it names. tsx shares the typescript node vocabulary, so those
    // two entries are the same string; the other three grammars spell their function forms
    // differently — python's `call` against everyone else's `call_expression` is the usual
    // example — which is what the per-language map exists for. The fingerprint is rooted at
    // the *body*, so within a language a method and a function of one shape still group.
    //
    // The `body:` field of `arrow_function` is `statement_block | expression`; matching the
    // block form alone is what keeps expression-bodied one-liners out. Python lambdas and
    // rust closures are excluded on the same reasoning: expression bodies are noise at any
    // threshold.
    query: {
      typescript: TS_QUERY,
      tsx: TS_QUERY,
      python: `
        (function_definition
          name: (identifier) @name
          body: (block) @body) @def
      `,
      go: `
        (function_declaration
          name: (identifier) @name
          body: (block) @body) @def

        (method_declaration
          name: (field_identifier) @name
          body: (block) @body) @def
      `,
      rust: `
        (function_item
          name: (identifier) @name
          body: (block) @body) @def
      `,
    },

    check(ctx, m) {
      const fp = ctx.structureFingerprint(m.body)
      if (!fp) return
      ctx.emitFact({
        kind: 'impl',
        hash: fp.hash,
        nodes: fp.nodes,
        line: ctx.line(m.def),
        column: ctx.column(m.def),
        name: m.name === undefined ? undefined : ctx.text(m.name),
      })
    },

    reduce(ctx) {
      const groups = new Map()
      for (const fact of ctx.facts('impl')) {
        if (fact.nodes < minNodes) continue
        const group = groups.get(fact.hash)
        if (group) {
          group.push(fact)
        } else {
          groups.set(fact.hash, [fact])
        }
      }

      for (const group of groups.values()) {
        if (group.length < 2) continue
        group.sort(byPosition)
        for (const member of group) {
          // `others` is the group minus this member, already sorted.
          const others = group.filter((fact) => fact !== member)
          ctx.report(
            { file: member.file, line: member.line, column: member.column },
            describe(member, others),
          )
        }
      }
    },
  })
}

/** The typescript grammar's function forms; tsx shares the vocabulary, so both entries use it. */
const TS_QUERY = `
  (function_declaration
    name: (identifier) @name
    body: (statement_block) @body) @def

  (method_definition
    body: (statement_block) @body) @def

  (function_expression
    body: (statement_block) @body) @def

  (arrow_function
    body: (statement_block) @body) @def
`

/** Facts in a stable order, so a message names the same counterparts every run. */
function byPosition(a, b) {
  if (a.file < b.file) return -1
  if (a.file > b.file) return 1
  return a.line - b.line || a.column - b.column
}

/**
 * The message for one member of a duplicate group.
 *
 * `others` is sorted; the first three are named, and the remainder collapsed into
 * "and N more" so the message stays bounded however large the group grows.
 */
function describe(member, others) {
  const shown = others.slice(0, 3)
  const list = shown.map((o) => `${o.file}:${o.line}`).join(', ')
  const rest = others.length - shown.length
  const capped = rest > 0 ? `${list} and ${rest} more` : list
  const prefix =
    member.name === undefined
      ? 'duplicated implementation — also at'
      : `'${member.name}' duplicates the implementation at`
  return `${prefix} ${capped}`
}

File: crates/lanekeep-cli/tests/duplicate_implementation.rs
//! `lanekeep/duplicate-implementation`, run through the real engine.
//!
//! A cross-file rule needs more than one file, so these build a project directly rather
//! than going through `RuleTester`, which runs one subject at a time.
//!
//! Fixture sizing matters and is deliberate: every body a test expects to group is well
//! above the default `minNodes` of 40 (measured in fingerprint nodes, which count kinds,
//! anonymous tokens and structure — a ten-line body is comfortably 50+), and every body a
//! test expects to stay silent is far below it. A "does not group" fixture is still above
//! the threshold, so its silence is proved to come from the shape difference and not from
//! being too small to participate.

#![expect(
    clippy::expect_used,
    clippy::panic,
    reason = "`clippy.toml`'s allow-*-in-tests only reaches `#[test]` functions and \
              `#[cfg(test)]` modules. The `corpus` helpers are neither, so the grant it \
              already makes for unit tests has to be restated for them."
)]

mod corpus;

use corpus::Corpus;

/// A body comfortably above the default `minNodes` threshold (≈ 60 fingerprint nodes).
const LARGE_FN: &str = "\
function compute(list) {
  const total = 0
  for (const item of list) {
    if (item.active) {
      total += item.value
    } else {
      total -= item.value
    }
  }
  return total
}
";

/// The same shape as [`LARGE_FN`] with every identifier and literal value changed — the
/// case the rule exists for.
const RENAMED_FN: &str = "\
function summarize(input) {
  const acc = 1
  for (const entry of input) {
    if (entry.ready) {
      acc += entry.score
    } else {
      acc -= entry.score
    }
  }
  return acc
}
";

/// A body far below the default threshold (≈ 8 fingerprint nodes).
const SMALL_FN: &str = "\
function tiny() {
  return 1
}
";

/// A class method and a block-bodied arrow, both above the threshold.
const METHODS_FN: &str = "\
class Widget {
  render(items) {
    const names = []
    const list = []
    for (const item of items) {
      if (item.enabled) {
        list.push(item.name)
      }
    }
    return list
  }
}

const summarize = (items) => {
  const list = []
  const seen = new Set()
  for (const item of items) {
    if (!seen.has(item)) {
      seen.add(item)
      list.push(item)
    }
  }
  return list
}
";

fn corpus(options: &str, files: &[(&str, &str)]) -> Corpus {
    Corpus::new("duplicate-implementation", options, files)
}

#[test]
fn two_identical_functions_in_different_files_report_both_sites() {
    let found = corpus("{}", &[("src/a.ts", LARGE_FN), ("src/b.ts", LARGE_FN)]).run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'compute' duplicates the implementation at src/b.ts:1",
            "src/b.ts:1:1 'compute' duplicates the implementation at src/a.ts:1",
        ]
    );
}

#[test]
fn renamed_identifiers_and_changed_literal_values_still_group() {
    // The point of the rule: `summarize` is `compute` with different names and numbers,
    // and an agent that cannot see the whole corpus has already written it.
    let found = corpus("{}", &[("src/a.ts", LARGE_FN), ("src/b.ts", RENAMED_FN)]).run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'compute' duplicates the implementation at src/b.ts:1",
            "src/b.ts:1:1 'summarize' duplicates the implementation at src/a.ts:1",
        ]
    );
}

#[test]
fn a_changed_operator_does_not_group() {
    // One `+=` becoming `-=` is a different body, whatever the two files look like side
    // by side. The fixture is above the threshold, so the silence is the operator's work.
    let b = LARGE_FN.replace("total += item.value", "total -= item.value");
    let found = corpus("{}", &[("src/a.ts", LARGE_FN), ("src/b.ts", &b)]).run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn an_added_statement_does_not_group() {
    // A body with an extra statement has a different shape. Same threshold argument as
    // the operator case: both bodies are big enough to participate, and do not.
    let b = LARGE_FN.replace(
        "  const total = 0\n",
        "  const total = 0\n  const extra = item.weight\n",
    );
    let found = corpus("{}", &[("src/a.ts", LARGE_FN), ("src/b.ts", &b)]).run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn a_comment_inside_the_body_does_not_hide_a_duplicate() {
    // Comments are erased before the body is fingerprinted, a docstring included — a
    // documented helper and its undocumented twin have the same shape and are flagged.
    // (A docstring is a comment in TypeScript, not a statement.)
    let b = LARGE_FN.replace(
        "  const total = 0\n",
        "  // total starts at zero\n  const total = 0\n",
    );
    let found = corpus("{}", &[("src/a.ts", LARGE_FN), ("src/b.ts", &b)]).run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'compute' duplicates the implementation at src/b.ts:1",
            "src/b.ts:1:1 'compute' duplicates the implementation at src/a.ts:1",
        ]
    );
}

#[test]
fn a_group_of_three_names_its_counterparts_in_a_stable_order() {
    // Every member names the other two, sorted by (file, line), so the messages are the
    // same on every run — an unsorted or emission-order list would flake under parallel
    // checking.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", LARGE_FN),
            ("src/b.ts", LARGE_FN),
            ("src/c.ts", LARGE_FN),
        ],
    )
    .run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'compute' duplicates the implementation at src/b.ts:1, src/c.ts:1",
            "src/b.ts:1:1 'compute' duplicates the implementation at src/a.ts:1, src/c.ts:1",
            "src/c.ts:1:1 'compute' duplicates the implementation at src/a.ts:1, src/b.ts:1",
        ]
    );
}

#[test]
fn many_counterparts_are_capped_in_the_message() {
    // A group of five names three counterparts and folds the rest, so the message stays
    // bounded however large the group grows.
    let found = corpus(
        "{}",
        &[
            ("src/a.ts", LARGE_FN),
            ("src/b.ts", LARGE_FN),
            ("src/c.ts", LARGE_FN),
            ("src/d.ts", LARGE_FN),
            ("src/e.ts", LARGE_FN),
        ],
    )
    .run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'compute' duplicates the implementation at src/b.ts:1, src/c.ts:1, src/d.ts:1 and 1 more",
            "src/b.ts:1:1 'compute' duplicates the implementation at src/a.ts:1, src/c.ts:1, src/d.ts:1 and 1 more",
            "src/c.ts:1:1 'compute' duplicates the implementation at src/a.ts:1, src/b.ts:1, src/d.ts:1 and 1 more",
            "src/d.ts:1:1 'compute' duplicates the implementation at src/a.ts:1, src/b.ts:1, src/c.ts:1 and 1 more",
            "src/e.ts:1:1 'compute' duplicates the implementation at src/a.ts:1, src/b.ts:1, src/c.ts:1 and 1 more",
        ]
    );
}

#[test]
fn small_bodies_stay_silent_below_the_default_threshold() {
    // Two identical two-line getters are noise, not a duplicate implementation. The
    // default of 40 keeps them out without anyone configuring anything.
    let found = corpus("{}", &[("src/a.ts", SMALL_FN), ("src/b.ts", SMALL_FN)]).run();
    assert_eq!(found, Vec::<String>::new());
}

#[test]
fn min_nodes_is_asserted_in_both_directions() {
    // Raising the threshold must silence a pair the default reports, and lowering it must
    // report a pair the default silences. A test that only lowered it would pass against
    // a rule that ignored the option.
    let lowered = corpus(
        "{ minNodes: 5 }",
        &[("src/a.ts", SMALL_FN), ("src/b.ts", SMALL_FN)],
    )
    .run();
    assert_eq!(
        lowered,
        vec![
            "src/a.ts:1:1 'tiny' duplicates the implementation at src/b.ts:1",
            "src/b.ts:1:1 'tiny' duplicates the implementation at src/a.ts:1",
        ]
    );

    let raised = corpus(
        "{ minNodes: 200 }",
        &[("src/a.ts", LARGE_FN), ("src/b.ts", LARGE_FN)],
    )
    .run();
    assert_eq!(raised, Vec::<String>::new());
}

#[test]
fn method_definitions_and_block_arrow_functions_are_covered() {
    // Function declarations are not the only way an implementation is written twice.
    // Methods and block-bodied arrows fingerprint the same way.
    let found = corpus("{}", &[("src/a.ts", METHODS_FN), ("src/b.ts", METHODS_FN)]).run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:2:3 duplicated implementation — also at src/b.ts:2",
            "src/a.ts:14:19 duplicated implementation — also at src/b.ts:14",
            "src/b.ts:2:3 duplicated implementation — also at src/a.ts:2",
            "src/b.ts:14:19 duplicated implementation — also at src/a.ts:14",
        ]
    );
}

#[test]
fn expression_bodied_arrows_are_not_matched() {
    // An expression-bodied one-liner is noise even when a matched body of the same size
    // would report — the threshold is lowered to 5 so a small block body does report,
    // which is what proves the expression pair's silence comes from the query, not from
    // being too small to participate.
    let file = "\
const pick = (x) => x.a.b.c + x.d.e.f
const block = (x) => { return x.a + x.b }
";
    let found = corpus("{ minNodes: 5 }", &[("src/a.ts", file), ("src/b.ts", file)]).run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:2:15 duplicated implementation — also at src/b.ts:2",
            "src/b.ts:2:15 duplicated implementation — also at src/a.ts:2",
        ]
    );
}

#[test]
fn two_identical_functions_in_one_file_are_reported() {
    // Same-file duplicates are duplicates too — the file-agnostic grouping reports every
    // member of a group, wherever the members sit.
    let mut both = String::from(LARGE_FN);
    both.push_str(&LARGE_FN.replace("function compute", "function second"));
    let found = corpus("{}", &[("src/a.ts", &both)]).run();
    assert_eq!(
        found,
        vec![
            "src/a.ts:1:1 'compute' duplicates the implementation at src/a.ts:12",
            "src/a.ts:12:1 'second' duplicates the implementation at src/a.ts:1",
        ]
    );
}

#[test]
fn the_same_corpus_reports_the_same_thing_every_run() {
    let corpus = corpus(
        "{}",
        &[
            ("src/a.ts", LARGE_FN),
            ("src/b.ts", RENAMED_FN),
            ("src/c.ts", LARGE_FN),
        ],
    );
    let first = corpus.run();
    assert!(!first.is_empty(), "the fixture should report something");
    for attempt in 0..4 {
        assert_eq!(corpus.run(), first, "output changed on attempt {attempt}");
    }
}
